Feature/update bam to cram

sb43 opened the pull request:
Updated bamToCram command line and samtools version in conda,

sb43 committed:
updated samtools version to 1.10 to enable indexing in a single command

sb43 committed:
updated samtools command to output index in a single command

sb43 committed:
updated samtools v1.10-0 and command in bamToCram to output bam index in a single command

bam to cram conversion now enabled for all bam files including qc  failed

sb43 committed:
bam to cram conversion now enabled for all bam files including qc  failed

sb43 commented:
Hi @sb10 please check the error, it seems missing sif files,  I have merged my master with upstream/master could that be an issue.

sb10 commented on workflows/articNcovNanopore.nf:
```diff
@@ -61,8 +61,8 @@ workflow sequenceAnalysisNanopolish {
      if (params.outCram) {
-        bamToCram(qc.pass.map{ it[0] } 
-                        .join (trimPrimerSequences.out.ptrim.combine(ch_preparedRef.map{ it[0] })) )
+        bamToCram(articMinIONNanopolish.out.ptirm.map{ it[0] } 
```
Is that a typo for ptrim? Should it be trimPrimerSequences.out.ptrim?

sb43 replied:
This is Nanopore workflow,  do I add bamToCram there or just keep it to Illumina?

sb10 replied:
I guess you haven't tested the nanopore workflow, but I think if you can have it work in theory it's worth doing it. I think we want to use nanopore workflow as well, but I don't have any test data.

sb43 replied:
Should work for nanopore bam as well just need to decide which bam files (mapped, primer trimmed or duplicate removed) need to convert

sb10 replied:
I'd say, convert the same bams that are used as input to QC.

sb43 replied:
yes, that's what I thought as well and used same input files as QC:
makeQCCSV(articMinIONNanopolish.out.ptrim,.....)

sb10 replied:
Ok, well it is still a typo. You've put ptirm, not ptrim.

sb43 replied:
thanks, corrected now

sb10 commented:
> Hi @sb10 please check the error, it seems missing sif files, I have merged my master with upstream/master could that be an issue.

I'm not sure if that's the issue. You can run what the github tests run locally to debug. It's `.github/scripts/test_PR_against_release.sh` that is failing, so just do what `.github/workflows/pull_request.yml` does before that test, then run it the same way.

sb43 commented:
> ,
Test continues even after failure in building containers.
bash scripts/build_singularity_containers.sh

Possibly due to issue in conda 4.8.2

Please see issue reported here:
https://github.com/conda/conda/issues/9681

+ /opt/conda/bin/conda env update -f /extras.yml -n artic-ncov2019
Collecting package metadata (repodata.json): done
Solving environment: | 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                                                                                                                                                            

# >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<

    Traceback (most recent call last):
      File "/opt/conda/lib/python3.7/site-packages/conda/exceptions.py", line 1079, in __call__
        return func(*args, **kwargs)
      File "/opt/conda/lib/python3.7/site-packages/conda_env/cli/main.py", line 80, in do_call
        exit_code = getattr(module, func_name)(args, parser)
      File "/opt/conda/lib/python3.7/site-packages/conda_env/cli/main_update.py", line 124, in execute
        result[installer_type] = installer.install(prefix, specs, args, env)
      File "/opt/conda/lib/python3.7/site-packages/conda_env/installers/conda.py", line 32, in install
        prune=getattr(args, 'prune', False), update_modifier=UpdateModifier.FREEZE_INSTALLED)
      File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 117, in solve_for_transaction
        should_retry_solve)
      File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 158, in solve_for_diff
        force_remove, should_retry_solve)
      File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 281, in solve_final_state
        ssc = self._run_sat(ssc)
      File "/opt/conda/lib/python3.7/site-packages/conda/common/io.py", line 88, in decorated
        return f(*args, **kwds)
      File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 808, in _run_sat
        should_retry_solve=ssc.should_retry_solve
      File "/opt/conda/lib/python3.7/site-packages/conda/common/io.py", line 88, in decorated
        return f(*args, **kwds)
      File "/opt/conda/lib/python3.7/site-packages/conda/resolve.py", line 1318, in solve
        self.find_conflicts(specs, specs_to_add, history_specs)
      File "/opt/conda/lib/python3.7/site-packages/conda/resolve.py", line 347, in find_conflicts
        bad_deps = self.build_conflict_map(specs, specs_to_add, history_specs)
      File "/opt/conda/lib/python3.7/site-packages/conda/resolve.py", line 507, in build_conflict_map
        root, search_node, dep_graph, num_occurances)
      File "/opt/conda/lib/python3.7/site-packages/conda/resolve.py", line 369, in breadth_first_search_for_dep_graph
        last_spec = MatchSpec.union((path[-1], target_paths[-1][-1]))[0]
      File "/opt/conda/lib/python3.7/site-packages/conda/models/match_spec.py", line 481, in union
        return cls.merge(match_specs, union=True)
      File "/opt/conda/lib/python3.7/site-packages/conda/models/match_spec.py", line 475, in merge
        reduce(lambda x, y: x._merge(y, union), group) if len(group) > 1 else group[0]
      File "/opt/conda/lib/python3.7/site-packages/conda/models/match_spec.py", line 475, in <lambda>
        reduce(lambda x, y: x._merge(y, union), group) if len(group) > 1 else group[0]
      File "/opt/conda/lib/python3.7/site-packages/conda/models/match_spec.py", line 502, in _merge
        final = this_component.union(that_component)
      File "/opt/conda/lib/python3.7/site-packages/conda/models/match_spec.py", line 764, in union
        return '|'.join(options)
    TypeError: sequence item 0: expected str instance, Channel found

`$ /opt/conda/bin/conda-env update -f /extras.yml -n artic-ncov2019`

  environment variables:
                 CIO_TEST=<not set>
  CONDA_AUTO_UPDATE_CONDA=false
               CONDA_ROOT=/opt/conda
                     PATH=/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin
       REQUESTS_CA_BUNDLE=<not set>
            SSL_CERT_FILE=<not set>

     active environment : None
       user config file : /root/.condarc
 populated config files : 
          conda version : 4.8.2
    conda-build version : not installed
         python version : 3.7.6.final.0
       virtual packages : __glibc=2.28
       base environment : /opt/conda  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /opt/conda/pkgs
                          /root/.conda/pkgs
       envs directories : /opt/conda/envs
                          /root/.conda/envs
               platform : linux-64
             user-agent : conda/4.8.2 requests/2.22.0 CPython/3.7.6 Linux/4.4.0-116-generic debian/10 glibc/2.28
                UID:GID : 0:0
             netrc file : None
           offline mode : False


An unexpected error has occurred. Conda has prepared the above report.

ABORT: Aborting with RETVAL=255
Cleaning up...
Built container /home/ubuntu/mymount2/heron/ncov2019-artic-nf/artic-ncov2019-illumina.sif
Built container /home/ubuntu/mymount2/heron/ncov2019-artic-nf/artic-ncov2019-nanopore.sif

sb10 commented:
Can you try with conda 4.8.3 and see if that fixes it?

sb43 commented:
> Can you try with conda 4.8.3 and see if that fixes it?

Doesn't  look straightforward as Dockerfile ( environments/illumina/Dockerfile)  uses prebuild 
FROM continuumio/miniconda3:latest which is 4.8.2

Also tried to upate conda after default installation, it doesn't seem to work:
RUN /opt/conda/bin/conda update -n base -c defaults conda

dkj committed:
Remove mpileup depth limit in consensus generation (#51)

The current samtools depth reduction technique is not suitable for the high depth NovaSeq output together with end of amplicon (ligated/tailed) adapters.

sb43 commented:
> > Can you try with conda 4.8.3 and see if that fixes it?
> 
> Doesn't look straightforward as Dockerfile ( environments/illumina/Dockerfile) uses prebuild
> FROM continuumio/miniconda3:latest which is 4.8.2
> 
> Also tried to upate conda after default installation, it doesn't seem to work:
> RUN /opt/conda/bin/conda update -n base -c defaults conda

@sb10 Shall I wait for upstream repo to fix the test for .sif generation ?

sb10 commented:
> > > Can you try with conda 4.8.3 and see if that fixes it?
> > 
> > 
> > Doesn't look straightforward as Dockerfile ( environments/illumina/Dockerfile) uses prebuild
> > FROM continuumio/miniconda3:latest which is 4.8.2
> > Also tried to upate conda after default installation, it doesn't seem to work:
> > RUN /opt/conda/bin/conda update -n base -c defaults conda
>
> @sb10 Shall I wait for upstream repo to fix the test for .sif generation ?

By upstream you mean connor-lab/ncov2019-artic-nf? No, if the test needs improving, just make and commit changes to the test as part of this PR.

Sorry, I don't understand the technicalities around conda, so can't help. If you can't get it working, just drop the "index at the same time" feature and limit this PR to just converting all the bams.

kjsanger commented:
Re. that Conda error, I was getting similar about 2 weeks ago, caused by upstream pinning Conda package versions in `environments/extras.yml` in such a way that the dependencies became unsolvable. Unfortunately Conda crashed on this in the same way as above so we couldn't see what the problem was.

Reverting Conda to the most recent `4.7.x` release and running the build again may tell you what's wrong in the Conda environment. It worked in my case.

(My rule of thumb is to avoid running a version of Conda more recent than the most patched _previous_ major release)

sb43 commented:
> Re. that Conda error, I was getting similar about 2 weeks ago, caused by upstream pinning Conda package versions in `environments/extras.yml` in such a way that the dependencies became unsolvable. Unfortunately Conda crashed on this in the same way as above so we couldn't see what the problem was.
> 
> Reverting Conda to the most recent `4.7.x` release and running the build again may tell you what's wrong in the Conda environment. It worked in my case.
> 
> (My rule of thumb is to avoid running a version of Conda more recent than the most patched _previous_ major release)

Thanks @kjsanger and @sb10,  I will try to revert conda to 4.7.x  and rerun the test

m-bull committed:
Update ivar consensus calling mpileup command (#54)

* Update ivar consensus calling mpileup command

Updated ivar documentation to adds `-aa` to `samtools mpileup` here: https://github.com/andersen-lab/ivar/blob/master/docs/MANUAL.md. cc @jts

* Move to Miniconda3 for tests

* Fix PR test script and temp allow tests to complete

m-bull committed:
Update tests to compare against  5e00ba9 (#55)

sb10 commented:
For the conda issue, try rebasing with latest upstream, which is using miniconda v3 for the tests.

sb43 committed:
Merge remote-tracking branch 'upstream/master' into feature/update_bam_to_cram

sb43 committed:
corrected typo

sb43 committed:
removed artefacts created during test

sb43 committed:
updated all conda to latest

m-bull committed:
Qc update (#56)

* Update qc.py to reflect changed standards

* Filter out Illumina Undetermined_S0

* Make sure to negate the matcher

kjsanger committed:
Add minReadsPerBarcode parameter for Nanopore workflow (#57)

The barcode<n*> directories of de-plexed Fastq input are filtered to
exclude any containing fewer than 5 files (guppy makes all of the
barcode directories, whether you used the barcodes or not).

This patch changes the filter to scan the files and count the records
within to 1) avoid excluding barcodes where many reads are in fewer
than 5 files and 2) avoid including barcodes where there are very few
reads in more than 5 files.

This threshold can be changed with the new optional parameter
minReadsPerBarcode, which defaults to 100 reads.

Co-authored-by: Keith James <dev@deoxybyte.co.uk>

ac55-sanger committed:
Remove unnecessary samtools view -bS (#48)

Co-authored-by: ac55-sanger <ac55@github.com>

dkj committed:
use ivar default of 30 for illuminaKeepLen, and comment (#61)

Co-authored-by: David K. Jackson <david.jackson@physics.org>

sb43 committed:
pulled changes from upstream master

sb43 committed:
pulled chnages from upstream master to run tests successfully

sb43 committed:
reverted changes from in the Docker and Singularity environments

sb43 committed:
added samtools v1.10 to allow cram file indexing in a single command

sb43 committed:
Merge remote-tracking branch 'origin/feature/update_bam_to_cram' into feature/update_bam_to_cram

sb10 reviewed (approved):
Thanks, is it all working now?

sb10 commented:
Can you see if you can resolve the conflicts to allow a rebase of your changes?

sb10 commented:
Or just cherry pick your changes in a new branch and do a PR direct to upstream.

sb43 commented:
> Or just cherry pick your changes in a new branch and do a PR direct to upstream.

I  did already have PR with upstream master, other commits from upstream master were merged with my branch, these are already part of connonr-lab master branch.
